### PR TITLE
Multi-user: use forking!

### LINKED ISSUE #4: Stop watching users that haven't bundled for a while
[`expire-old-webpacks`](https://github.com/iFixit/multi-user-dev-server/compare/expire-old-webpacks)

`multi-user-dev-server` eats boattons of memory. We believe this is because we have lots and lots of watches running simultaneously. We should drop watches for users after some short terms worth of inactivity.

Do something like this
---

See:
https://github.com/iFixit/multi-user-dev-server/blob/47d4009ac7460117a7007ed08cd7db5f866c5b03/multi-user-dev-server.js#L72
Let's keep the timestamp of the most recent action performed by a user.

Then we can `setInterval` and clear away watches for users if they have not accessed in some `X` number of minutes.

We can borrow the logic for closing a watch from  here:
https://github.com/iFixit/multi-user-dev-server/blob/47d4009ac7460117a7007ed08cd7db5f866c5b03/multi-user-dev-server.js#L94-L95

Ref: https://webpack.js.org/api/node/

Hopefully this will decrease the severe memory usage on the dev server.

CC @danielbeardsley

### PR DESCRIPTION
**Pull request by danielbeardsley**
Turns out, webpack uses tons of memory and doesn't seem to release
it after you call watching.close() and null out the reference.

Not exactly sure what's up with that, but process isolation will
save us.

We now run a node process for each user whose webpack dir we want to
watch. We use node's IPC to send messages back and forth between
parent and child.

Noteable things:
* Collection of promises wating for a build has been moved
  into webpack-watcher
* Compiler expiration now kills the whole forked process,
  fully releasing the heap it was using

Closes #4

### CODE REVIEW

File: compiler-collection.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+modules.exports = () => {
+  const compilers = {}
```
Comment (djmetzle) [inline compiler-collection.js]: This collection might be best represented as a `Map`? Just a thought.

File: multi-user-dev-server.js [hunk 2]
```diff
@@ -24,56 +24,38 @@ const CompilerCollection = require("./compiler-collection.js");
...
-   * @return { compiler, watching, whenDone }
+   * @return { compiler, watching }
    */
   const getUserCompiler = username => {
     if (!/^[\w\-_]+$/.test(username)) {
       throw new Error('invalid characters in username');
     }
 
     const options = optionsFromUsername(username);
+    options.username = username;
```
Comment (djmetzle) [inline multi-user-dev-server.js:48]: Wat? That seems weird. I would think that `optionsFromUsername` would add this value to the `options` object?

File: multi-user-dev-server.js [hunk 3]
```diff
@@ -24,56 +24,38 @@ const CompilerCollection = require("./compiler-collection.js");
...
     // user's webpack config, since it may have changed since this server
     // started.
-    delete require.cache[require.resolve(options.configPath)];
-    const getWebpackConfig = require(options.configPath);
-    const compiler = WebpackWatcher(getWebpackConfig(options.webpackEnv || {}));
-
-    // `whenDone` will add pending promises to this list, which will be resolved
-    // when the `watching` handler gets called.
-    let promises = [];
+    const compiler = WebpackWatcher(options);
```
Comment (djmetzle) [inline multi-user-dev-server.js:50]: The comment here seems irrelevant now?

File: webpack-watcher.js [hunk 4]
```diff
@@ -1,2 +1,58 @@
...
+      });
+   }
+
+   const watcher = {
+      close: () => {
+         console.log("Killing " + username);
+         child.kill();
+         notifyWatchers()
+      },
+      whenDone
```
Comment (djmetzle) [inline webpack-watcher.js]: Just `whenDone`? Is this a special syntax thing? It seems erroneous.

**Comment by djmetzle** [comment]
This is really awesome @danielbeardsley!

I do have some critique though, that i hope doesn't come off as too pushy. I think that maybe we should use ES6 classes in a few places. The [watcher object](https://github.com/iFixit/multi-user-dev-server/blob/139ba8a90827de9d833a91981828cb3cf926084f/webpack-watcher.js) seems to be a good candidate for a object proper in it's own right.

Also, i think there is some fork/module/IPC magic going on in the [compiler include file](https://github.com/iFixit/multi-user-dev-server/blob/139ba8a90827de9d833a91981828cb3cf926084f/webpack-compiler.js) that could be separated out into it's own thing. I would like to see a `compiler-forker-entrypoint` module for just kicking off the child pid. I think that separating out the entrypoint could help to understand and follow the logical flow.

Just some pesky structural comments. I hope they're helpful. Feel free to disregard.

CR :medal_military:

File: webpack-watcher.js (hunk 4 above)
Comment (sctice) [reply webpack-watcher.js]: It creates a property named `whenDone` with the value of the `whenDone` variable. It's a shorthand for constructing objects from variables.

I don't really see why we're defining it this way instead of in the same way as `close`. Am I missing a use of `whenDone` elsewhere?

File: compiler-collection.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+module.exports = (expireAfterSeconds) => {
+  const expireMs = expireAfterSeconds * 1000;
+  const compilers = new Map();
+
+  if (expireAfterSeconds) {
+    setInterval(removeExpiredCompilers, 30 * 1000);
```
Comment (sctice) [inline compiler-collection.js]: It seems to me that this interval should be based on `expireAfterSeconds`. Otherwise, an `expireAfterSeconds < 30` doesn't really do what it advertises.

File: compiler-collection.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    return compiler.lastAccessed < oldestAllowedAccessTime;
+  }
+
+  function updateLastAccessed(username) {
+    const compiler = compilers.get(username)
+    if (!compiler) return;
+    compiler.lastAccessed = Date.now();
+  }
+
+  const interface = {
```
Comment (sctice) [inline compiler-collection.js]: `interface` is a [future reserved word](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Future_reserved_keywords). I think it would be a good idea to use something else, maybe `collection`?

File: webpack-compiler.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+         if (!watcher) {
+            watcher = getWebpack(message.options, message.watchOptions);
+         }
+         break;
+
+      // The parent is asking if we're in the middle of running (building)
+      case 'isRunning':
+         // If we're not running (building), let them know we're done;
+         // otherwise, they'll find out via the 'built' event.
+         if (!watcher.running) {
```
Comment (sctice) [inline webpack-compiler.js]: I think this is assuming a well-behaved parent, who never sends `isRunning` before `watch`.

File: webpack-watcher.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+         watchers.push({resolve, reject});
+         child.send({event: 'isRunning'})
+      });
+   }
+
+   const watcher = {
+      close: () => {
+         console.log("Killing " + options.username);
+         child.kill();
+         notifyWatchers()
```
Comment (sctice) [inline webpack-watcher.js]: It strikes me as odd that this tells the watchers that bundling succeeded.

File: webpack-watcher.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+         const success = err ? " has failed" : " has succeeded";
+         console.log("notifying " + watchers.length + " watchers that bundling " + options.username + success);
+      }
+      watchers.forEach(({resolve, reject}) => err ? reject(err) : resolve());
+      watchers = [];
+   }
+
+   function whenDone() {
+      return new Promise((resolve, reject) => {
+         watchers.push({resolve, reject});
```
Comment (sctice) [inline webpack-watcher.js]: I think `watchers` and `notifyWatchers` are misleading names here when we have a thing called `watcher`, and that's _not_ what goes into `watchers`. Maybe something like `whenDoneObservers` or `whenDonePromiseResolvers`?

File: webpack-watcher.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+const childProcess = require('child_process');
+
+module.exports = (options) => {
+   console.log("Forking " + options.username);
+   const child = childProcess.fork(__dirname + '/webpack-compiler.js')
+
+   let bundledCallback = () => {}
```
Comment (sctice) [inline webpack-watcher.js]: I find this name a little unintuitive because my first reading is to treat "bundled" as an adjective, i.e., a callback that is somehow "bundled". Since it's effectively an event handler, I'd suggest something like `onBundled`.

File: webpack-watcher.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+   let watchers = [];
```
Comment (sctice) [inline webpack-watcher.js]: This could be a `const` if we emptied the list of watchers instead of replacing it. I think that would be a little more true to the way it's used. Unfortunately, there isn't a super intuitive way to empty an array (unless you think `watchers.length = 0` is intuitive), but if we replaced this with a `Set`, then we could do `watchers.clear()`.

**Comment by sctice** [comment]
Cool stuff. I agree with @djmetzle's comments about classes, and I left a few other small suggestions and a couples questions about what may be bugs worth fixing. CR 🎏 but deploy_block ✋ on questions.

File: multi-user-dev-server.js (hunk 2 above)
Comment (danielbeardsley) [reply multi-user-dev-server.js:48]: Yes, but this function is outside the app (i.e. the mapping function is part of the input to the app) so placing an arbitrary restriction on it (that we could easily fulfill ourselves) seemed like a bad idea.

File: multi-user-dev-server.js (hunk 3 above)
Comment (danielbeardsley) [reply multi-user-dev-server.js:50]: Removing!

File: multi-user-dev-server.js (hunk 2 above)
Comment (sctice) [reply multi-user-dev-server.js:48]: Fair enough. That might be worth a comment since it looked odd to both of us.

File: webpack-compiler.js (hunk 7 above)
Comment (danielbeardsley) [reply webpack-compiler.js]: Good call, easy to address.

File: webpack-watcher.js (hunk 8 above)
Comment (danielbeardsley) [reply webpack-watcher.js]: Fair. I'll make it reject!

File: webpack-watcher.js (hunk 4 above)
Comment (danielbeardsley) [reply webpack-watcher.js]: Moved it inside!

File: compiler-collection.js (hunk 5 above)
Comment (danielbeardsley) [reply compiler-collection.js]: True. But, to be fair, an expiration of 30s or less is probably a bad idea. Either way, I'll make the change.

File: compiler-collection.js (hunk 6 above)
Comment (danielbeardsley) [reply compiler-collection.js]: TIL! Fixing!

File: webpack-watcher.js (hunk 9 above)
Comment (danielbeardsley) [reply webpack-watcher.js]: I went for renaming the public interface instead.

File: compiler-collection.js (hunk 5 above)
Comment (sctice) [reply compiler-collection.js]: > True. But, to be fair, an expiration of 30s or less is probably a bad idea.

This coming from the guy who [made the example 10](https://github.com/iFixit/multi-user-dev-server/pull/6/files#diff-e55bc98107f12615bba2a6649f130228R12) 😁.

**Comment by djmetzle** [comment]
CR :fork_and_knife:

**Comment by danielbeardsley** [comment]
> Also, i think there is some fork/module/IPC magic going on in the compiler include file that could be separated out into it's own thing

You mean the message sending and receiving? I'm not sure what parts I would pull out of [this file](https://github.com/iFixit/multi-user-dev-server/blob/139ba8a90827de9d833a91981828cb3cf926084f/webpack-compiler.js) to make the code more clear.

I'm confused about this:

>  I would like to see a compiler-forker-entrypoint module for just kicking off the child pid.

But an entrypoint would be the thing you would run in the subprocess, thus it wouldn't do the "kicking off the child pid", that'd be the job of the parent.

**Comment by danielbeardsley** [comment]
Thanks for all the suggestions! I liked them all.

un_deploy_block 👍  There's definitely more options for refactoring here, but I don't want to rewrite it entirely, so I'm ok with leaving the code that doesn't need to be touched as-is.

File: compiler-collection.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
+module.exports = (expireAfterSeconds) => {
+  const expireMs = expireAfterSeconds * 1000;
+  const compilers = new Map();
+
+  if (expireAfterSeconds) {
+    // Check for expired things often enough to expire things within
+    // 20% of their true expiration time.
+    setInterval(removeExpiredCompilers, expireAfterSeconds / 5);
```
Comment (sctice) [inline compiler-collection.js]: It occurs to me that a simpler way to do this might be to register a per-compiler timeout whenever we would `updateLastAccessed` for `expireMs` and cancel any existing timeout at the same time. I think it's basically as easy as storing the last accessed time, and then we expire things exactly when we expect to. Does that make sense?

File: multi-user-dev-server.js [hunk 13]
```diff
@@ -15,61 +16,46 @@ const Express = require("express");
  * @param optionsFromUsername A function that takes a username and returns a
  *                            options for multi-user-dev-server. See sample in
  *                            example/app.js.
+ * @param expireUnusedAfterSeconds Number of seconds after which webpack
+ *                            watchers that are unused will be closed /
+ *                            released. 0 or null == no expiration.
```
Comment (sctice) [inline multi-user-dev-server.js:21]: I am compelled to comment that this style of indentation is silly, and I don't understand why some people prefer it. Indentation should be consistent so that the amount of whitespace tells us something about nesting rather than depending purely on the size of a name!

Of course, this isn't directed at you. You're just correctly matching style.

File: webpack-watcher.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+   const watchController = {
+      close: () => {
+         console.log("Killing " + options.username);
+         child.kill();
+         notifyWatchers("Webpack killed before building completed")
+      },
+      whenDone: () => new Promise((resolve, reject) => {
+         watchers.add({resolve, reject});
+         child.send({event: 'isRunning'})
+         console.log("asking isRunning");
```
Comment (sctice) [inline webpack-watcher.js]: Something feels weird to me about the `isRunning` message, but I'm having trouble putting my finger on it. It seems somehow superfluous or mismatched to what we're trying to accomplish.

Could we just maintain our own internal state for "asked for build" and "build done" and refer to that in `whenDone` to decide whether to `notifyWatchers()`?

**Comment by sctice** [comment]
CR 🎏 but deploy_block ✋ on considering a couple more suggestions.

File: webpack-watcher.js (hunk 14 above)
Comment (danielbeardsley) [reply webpack-watcher.js]: The problem is there is no "build started" event that we can listen for to know the state, all we can do is ask "is running". We could poll it fast and let the parent know when it changes, but that's just more code. I'm open to it, but I'm not sure it's a better solution

File: compiler-collection.js (hunk 12 above)
Comment (danielbeardsley) [reply compiler-collection.js]: Not a bad idea. I think that's probably a bit more intuitive and probably less code. Doing it!

File: webpack-watcher.js (hunk 14 above)
Comment (sctice) [reply webpack-watcher.js]: Ah, I see. We know when we ask it to start watching, and we know when it finishes building something, but it doesn't tell us when it's started building something. That's a bummer.

**Comment by danielbeardsley** [comment]
un_deploy_block 👍  followed your suggestions!

File: compiler-collection.js [hunk 15]
```diff
@@ -2,40 +2,24 @@ module.exports = (expireAfterSeconds) => {
...
-
-  function updateLastAccessed(username) {
+  function resetExpireTimeout(username) {
     const compiler = compilers.get(username)
-    if (!compiler) return;
-    compiler.lastAccessed = Date.now();
+    if (!compiler || !expireMs) {
+       return;
+    }
+    clearTimeout(compiler.expiredTimeout);
```
Comment (djmetzle) [inline compiler-collection.js:10]: Neat.

**Comment by djmetzle** [comment]
More CR :gear:

**Comment by sctice** [comment]
Sweet! CR 🎏

**Comment by djmetzle** [comment]
CR :stopwatch:

**Comment by sctice** [comment]
Whoops! Totally missed that. CR 🎏

**Comment by danielbeardsley** [comment]
This has been running fine for a few days. I'm gonna merge it.
